### ISSUE TITLE
add support for inline templates

### ISSUE THREAD

**Issue by monperrus** [INRIA/spoon#3118]
inline templates:

  * with anonymous classes
  * with lambda

**Comment by monperrus** [comment]
tentatively adding label good-first-issue, but maybe it's too hard :-)

**Comment by LakshyAAAgrawal** [comment]
Hey, could someone please give me a little background on this so that I could work on it? Or should I give some other (easier?) issue a try before this and come back to it later?

**Comment by monperrus** [comment]
Great! Clarified the intent in the appropriate documentation file about Spoon templates, see #3516

**Comment by Rohitesh-Kumar-Jain** [comment]
Hi @monperrus,

I am confused, is one supposed to rewrite an existing method as an inline method using lambda/anonymous classes inside another method??

**Comment by monperrus** [comment]
No, the goal is to support writing templates with inline classes (which is not possible today), see #3516.

**Comment by KishkinJ10** [comment]
Sir in GSoC coding period students have to fix issues and add some features ??

**Comment by LakshyAAAgrawal** [comment]
Hi Everyone,

This is a really interesting issue and I have tried to understand how to solve this. Kindly correct me if I am wrong in any of the following observations:
1. The spoon template support needs access to the source code of the template class to generate the templates.
2. Spoon can be informed about the location of the template source file using ```Launcher.addTemplateResource(..)```
3. In case of anonymous classes and Lambdas, the source for the template class may reside on client source file(outside of spoon directory, and in the directory of caller), in which case the caller can simply call ```Launcher.addTemplateResource(<path to source of calling class>)``` and the inline-(anonymous or lambda) template works as expected. A working and tested example of this can be found at https://github.com/INRIA/spoon/compare/master...LakshyAAAgrawal:inlinetemplateaddSPexample. 
4. The Java Reflections API provides a lot of information at runtime however, it cannot provide information about the source within Methods
5. If we do not want the user to have to specify the source location of the class in which the anonymous template is defined(as in step 3), then the other options to access the source code could be
    1. Have the user pass the transformation as a string field in the anonymous class, which is accessed via reflection in ```spoon.template.Substitution#getTemplateCtClass```, and added into the spoon model dynamically. This however is prone to IllegalAccessException in the presence of a Security Manager, in which case even ```java.lang.reflect.AccessibleObject.setAccessible(true)``` fails. And there is no way to change access modifier of an anonymous class.
    2. Have the user pass the full source of the template as a string itself. But I feel this defeats the whole purpose of having anonymous classes.
    3. Consider using a decompiler to decompile the class. This is most likely to fail imo.

At the moment, I could come up with these observations, but I believe a much simpler solution should exist, and with a little bit of guidance, I would be able to find it. Kindly let me know what do you think.

**Comment by monperrus** [comment]
Thanks for the analysis.

For in-line templates, the whole source code of the transformations containing the inline templates has to be passed to `addTemplateResource` (as opposed to only the template classes).

**Comment by LakshyAAAgrawal** [comment]
Sir,

Could you please clarify what is the actionable point to close this issue? As per my understanding, a documentation task is remaining to clarify point number 1 and 3 from my previous comment in the spoon docs. Kindly let me know if there is any other task remaining.
